File: app/src/main/java/remix/myplayer/misc/menu/AudioPopupListener.java
package remix.myplayer.misc.menu;

import android.content.ContextWrapper;
import android.content.Intent;
import android.media.audiofx.AudioEffect;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.PopupMenu;
import android.view.MenuItem;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import remix.myplayer.App;
import remix.myplayer.Global;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.PlayListSong;
import remix.myplayer.bean.mp3.Song;
import remix.myplayer.helper.MusicServiceRemote;
import remix.myplayer.service.Command;
import remix.myplayer.service.MusicService;
import remix.myplayer.ui.Tag;
import remix.myplayer.ui.activity.EQActivity;
import remix.myplayer.ui.activity.PlayerActivity;
import remix.myplayer.ui.dialog.AddtoPlayListDialog;
import remix.myplayer.ui.dialog.FileChooserDialog;
import remix.myplayer.ui.dialog.TimerDialog;
import remix.myplayer.ui.fragment.LyricFragment;
import remix.myplayer.util.Constants;
import remix.myplayer.util.MediaStoreUtil;
import remix.myplayer.util.PlayListUtil;
import remix.myplayer.util.SPUtil;
import remix.myplayer.util.ToastUtil;
import remix.myplayer.util.Util;

import static com.afollestad.materialdialogs.DialogAction.POSITIVE;
import static remix.myplayer.helper.MusicServiceRemote.getCurrentSong;
import static remix.myplayer.helper.MusicServiceRemote.getMediaPlayer;
import static remix.myplayer.theme.Theme.getBaseDialog;
import static remix.myplayer.util.Util.sendLocalBroadcast;

/**
 * @ClassName AudioPopupListener
 * @Description
 * @Author Xiaoborui
 * @Date 2016/8/29 15:33
 */
public class AudioPopupListener<ActivityCallback extends AppCompatActivity & FileChooserDialog.FileCallback> extends ContextWrapper implements PopupMenu.OnMenuItemClickListener {
    private ActivityCallback mActivity;
    private Song mInfo;
    private Tag mTag;

    public AudioPopupListener(ActivityCallback activity, Song song) {
        super(activity);
        mActivity = activity;
        mInfo = song;
        mTag = new Tag(activity, song);
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_lyric:
                final boolean alreadyIgnore = SPUtil.getValue(mActivity, SPUtil.LYRIC_KEY.NAME, mInfo.getId() + "", SPUtil.LYRIC_KEY.LYRIC_DEFAULT) == SPUtil.LYRIC_KEY.LYRIC_IGNORE;
                final LyricFragment lyricFragment = ((PlayerActivity) mActivity).getLyricFragment();
                getBaseDialog(mActivity)
                        .items(getString(R.string.netease),
                                getString(R.string.kugou),
                                getString(R.string.select_lrc),
                                getString(!alreadyIgnore ? R.string.ignore_lrc : R.string.cancel_ignore_lrc),
                                getString(R.string.change_offset))
                        .itemsColorAttr(R.attr.text_color_primary)
                        .backgroundColorAttr(R.attr.background_color_3)
                        .itemsCallback((dialog, itemView, position, text) -> {
                            switch (position) {
                                case 0: //网易 酷狗
                                case 1:
                                    SPUtil.putValue(mActivity, SPUtil.LYRIC_KEY.NAME, mInfo.getId() + "", position == 0 ? SPUtil.LYRIC_KEY.LYRIC_NETEASE : SPUtil.LYRIC_KEY.LYRIC_KUGOU);
                                    lyricFragment.updateLrc(mInfo, true);
                                    sendLocalBroadcast(new Intent(MusicService.ACTION_CMD).putExtra("Control", Command.CHANGE_LYRIC));
                                    break;
                                case 2: //手动选择歌词
                                    new FileChooserDialog.Builder(mActivity)
                                            .extensionsFilter(".lrc")
                                            .show();
                                    break;
                                case 3: //忽略或者取消忽略
                                    getBaseDialog(mActivity)
                                            .title(!alreadyIgnore ? R.string.confirm_ignore_lrc : R.string.confirm_cancel_ignore_lrc)
                                            .negativeText(R.string.cancel)
                                            .positiveText(R.string.confirm)
                                            .onPositive((dialog1, which) -> {
                                                if (!alreadyIgnore) {//忽略
                                                    if (mInfo != null) {
                                                        SPUtil.putValue(mActivity, SPUtil.LYRIC_KEY.NAME, mInfo.getId() + "", SPUtil.LYRIC_KEY.LYRIC_IGNORE);
                                                        lyricFragment.updateLrc(mInfo);
                                                    }
                                                } else {//取消忽略
                                                    SPUtil.putValue(mActivity, SPUtil.LYRIC_KEY.NAME, mInfo.getId() + "", SPUtil.LYRIC_KEY.LYRIC_DEFAULT);
                                                    lyricFragment.updateLrc(mInfo);
                                                }
                                                sendLocalBroadcast(new Intent(MusicService.ACTION_CMD).putExtra("Control", Command.CHANGE_LYRIC));
                                            })
                                            .show();
                                    break;
                                case 4://歌词时间轴调整
                                    ((PlayerActivity) mActivity).showLyricOffsetView();
                                    break;
                            }

                        })
                        .show();
                break;
            case R.id.menu_edit:
                mTag.edit();
                break;
            case R.id.menu_detail:
                mTag.detail();
                break;
            case R.id.menu_timer:
                mActivity.startActivity(new Intent(mActivity, TimerDialog.class));
                break;
            case R.id.menu_eq:
                Intent audioEffectIntent = new Intent(AudioEffect.ACTION_DISPLAY_AUDIO_EFFECT_CONTROL_PANEL);
                audioEffectIntent.putExtra(AudioEffect.EXTRA_AUDIO_SESSION, getMediaPlayer().getAudioSessionId());
                if (Util.isIntentAvailable(mActivity, audioEffectIntent)) {
                    mActivity.startActivityForResult(audioEffectIntent, 0);
                } else {
                    mActivity.startActivity(new Intent(mActivity, EQActivity.class));
                }
                break;
            case R.id.menu_collect:
                PlayListSong info = new PlayListSong(mInfo.getId(), Global.MyLoveID, Constants.MYLOVE);
                ToastUtil.show(mActivity,
                        PlayListUtil.addSong(info) > 0 ? getString(R.string.add_song_playlist_success, 1, Constants.MYLOVE) : getString(R.string.add_song_playlist_error));
                break;
            case R.id.menu_add_to_playlist:
                Intent intentAdd = new Intent(this, AddtoPlayListDialog.class);
                Bundle ardAdd = new Bundle();
                ardAdd.putSerializable("list", new ArrayList<>(Collections.singletonList(mInfo.getId())));
                intentAdd.putExtras(ardAdd);
                startActivity(intentAdd);
                break;
            case R.id.menu_delete:
                getBaseDialog(mActivity)
                        .content(R.string.confirm_delete_from_library)
                        .positiveText(R.string.confirm)
                        .negativeText(R.string.cancel)
                        .checkBoxPromptRes(R.string.delete_source, SPUtil.getValue(App.getContext(), SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.DELETE_SOURCE, false), null)
                        .onAny((dialog, which) -> {
                            if (which == POSITIVE) {
                                if (MediaStoreUtil.delete(mInfo.getId(), Constants.SONG, dialog.isPromptCheckBoxChecked()) > 0) {
                                    if (PlayListUtil.deleteSong(mInfo.getId(), Global.PlayQueueID)) {
                                        ToastUtil.show(mActivity, getString(R.string.delete_success));
                                        //移除的是正在播放的歌曲
                                        if (getCurrentSong() == null)
                                            return;
                                        if (mInfo.getId() == getCurrentSong().getId()) {
                                            Intent intent = new Intent(MusicService.ACTION_CMD);
                                            intent.putExtra("Control", Command.NEXT);
                                            Util.sendLocalBroadcast(intent);
                                        }
                                    }
                                } else {
                                    ToastUtil.show(mActivity, getString(R.string.delete_error));
                                }
                            }
                        })
                        .show();
                break;
//            case R.id.menu_vol:
//                AudioManager audioManager = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
//                if(audioManager != null){
//                    audioManager.setStreamVolume(AudioManager.STREAM_MUSIC,audioManager.getStreamVolume(AudioManager.STREAM_MUSIC),AudioManager.FLAG_PLAY_SOUND | AudioManager.FLAG_SHOW_UI);
//                }
            case R.id.menu_speed:
                final List<String> speeds = Arrays.asList("0.5", "0.75", "1.0", "1.25", "1.5");
                final String originalSpeed = SPUtil.getValue(mActivity, SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.SPEED, "1.0");
                getBaseDialog(mActivity)
                        .title(R.string.speed)
                        .items(speeds)
                        .itemsCallbackSingleChoice(speeds.indexOf(originalSpeed), (dialog, itemView, which, text) -> {
                            MusicServiceRemote.setSpeed(Float.parseFloat(text.toString()));
                            SPUtil.putValue(mActivity, SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.SPEED, text.toString());
                            return true;
                        }).show();
                break;
        }
        return true;
    }

}

File: app/src/main/java/remix/myplayer/misc/menu/AudioPopupListener.kt
package remix.myplayer.misc.menu

import android.content.ActivityNotFoundException
import android.content.ContextWrapper
import android.view.MenuItem
import android.webkit.MimeTypeMap
import android.widget.CompoundButton
import androidx.appcompat.widget.PopupMenu
import com.afollestad.materialdialogs.DialogAction.POSITIVE
import com.afollestad.materialdialogs.MaterialDialog
import remix.myplayer.R
import remix.myplayer.bean.mp3.Song
import remix.myplayer.db.room.DatabaseRepository
import remix.myplayer.helper.DeleteHelper
import remix.myplayer.helper.EQHelper
import remix.myplayer.helper.LyricsHelper
import remix.myplayer.helper.MusicServiceRemote
import remix.myplayer.helper.MusicServiceRemote.getCurrentSong
import remix.myplayer.lyrics.provider.EmbeddedProvider
import remix.myplayer.lyrics.provider.IgnoredProvider
import remix.myplayer.lyrics.provider.StubProvider
import remix.myplayer.service.Command
import remix.myplayer.theme.Theme.getBaseDialog
import remix.myplayer.ui.activity.PlayerActivity
import remix.myplayer.ui.dialog.AddtoPlayListDialog
import remix.myplayer.ui.dialog.TimerDialog
import remix.myplayer.ui.misc.AudioTag
import remix.myplayer.util.RxUtil.applySingleScheduler
import remix.myplayer.util.SPUtil
import remix.myplayer.util.ToastUtil
import remix.myplayer.util.Util
import java.lang.ref.WeakReference

/**
 * @ClassName AudioPopupListener
 * @Description
 * @Author Xiaoborui
 * @Date 2016/8/29 15:33
 */
class AudioPopupListener(activity: PlayerActivity, private val song: Song) :
    ContextWrapper(activity), PopupMenu.OnMenuItemClickListener {
  private val audioTag: AudioTag = AudioTag(activity, song)
  private val ref = WeakReference(activity)

  override fun onMenuItemClick(item: MenuItem): Boolean {
    val activity = ref.get() ?: return true
    when (item.itemId) {
      R.id.menu_lyric -> {
        LyricsHelper.showLocalLyricsTip(activity) {
          onClickLyric(activity)
        }
        return true
      }
      R.id.menu_edit -> {
        if (!song.isLocal()) {
          return true
        }
        audioTag.edit()
      }
      R.id.menu_detail -> {
        audioTag.detail()
      }
      R.id.menu_timer -> {
        val fm = activity.supportFragmentManager
        TimerDialog.newInstance().show(fm, TimerDialog::class.java.simpleName)
      }
      R.id.menu_eq -> {
        EQHelper.startEqualizer(activity)
      }
      R.id.menu_collect -> {
        if (!song.isLocal()) {
          return true
        }
        DatabaseRepository.getInstance()
            .insertToPlayList(listOf(song.id), getString(R.string.my_favorite))
            .compose(applySingleScheduler())
            .subscribe(
                { count -> ToastUtil.show(activity, getString(R.string.add_song_playlist_success, 1, getString(R.string.my_favorite))) },
                { throwable -> ToastUtil.show(activity, R.string.add_song_playlist_error) })
      }
      R.id.menu_add_to_playlist -> {
        if (!song.isLocal()) {
          return true
        }
        AddtoPlayListDialog.newInstance(listOf(song.id))
            .show(activity.supportFragmentManager, AddtoPlayListDialog::class.java.simpleName)
      }
      R.id.menu_delete -> {
        if (!song.isLocal()) {
          return true
        }
        val checked = arrayOf(SPUtil.getValue(activity, SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.DELETE_SOURCE, false))

        getBaseDialog(activity)
            .content(R.string.confirm_delete_from_library)
            .positiveText(R.string.confirm)
            .negativeText(R.string.cancel)
            .checkBoxPromptRes(R.string.delete_source, checked[0], CompoundButton.OnCheckedChangeListener { buttonView, isChecked -> checked[0] = isChecked })
            .onAny { dialog, which ->
              if (which == POSITIVE) {
                DeleteHelper.deleteSong(activity, song.id, checked[0], false, "")
                    .compose<Boolean>(applySingleScheduler<Boolean>())
                    .subscribe({ success ->
                      if (success) {
                        //移除的是正在播放的歌曲
                        if (song.id == getCurrentSong().id) {
                          Util.sendCMDLocalBroadcast(Command.NEXT)
                        }
                      }
                      ToastUtil.show(activity, if (success) R.string.delete_success else R.string.delete_error)
                    }, { ToastUtil.show(activity, R.string.delete) })
              }
            }
            .show()
      }
      //            case R.id.menu_vol:
      //                AudioManager audioManager = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
      //                if(audioManager != null){
      //                    audioManager.setStreamVolume(AudioManager.STREAM_MUSIC,audioManager.getStreamVolume(AudioManager.STREAM_MUSIC),AudioManager.FLAG_PLAY_SOUND | AudioManager.FLAG_SHOW_UI);
      //                }
      R.id.menu_speed -> {
        getBaseDialog(activity)
            .title(R.string.speed)
            .input(SPUtil.getValue(activity, SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.SPEED, "1.0"),
                "",
                MaterialDialog.InputCallback { dialog, input ->
                  var speed = 0f
                  try {
                    speed = java.lang.Float.parseFloat(input.toString())
                  } catch (ignored: Exception) {

                  }

                  if (speed > 2f || speed < 0.5f) {
                    ToastUtil.show(activity, R.string.speed_range_tip)
                    return@InputCallback
                  }
                  SPUtil.putValue(activity, SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.SPEED,
                      input.toString())
                })
            .show()
      }
    }
    return true
  }

  private fun onClickLyric(activity: PlayerActivity) {
    getBaseDialog(activity).items(
        getString(R.string.default_lyrics),
        getString(R.string.embedded_lyric),
        getString(R.string.local),
        getString(R.string.kugou),
        getString(R.string.netease),
        getString(R.string.qq),
        getString(R.string.select_lrc),
        getString(R.string.ignore_lrc),
        getString(R.string.lyric_adjust_font_size),
        getString(R.string.change_offset)
      ).itemsCallback { _, _, position, _ ->
        when (position) {
          0 -> MusicServiceRemote.service?.updateLyrics(StubProvider) // 恢复默认
          1 -> MusicServiceRemote.service?.updateLyrics(EmbeddedProvider) // 内嵌
          2, 3, 4, 5 -> TODO() //  本地 酷狗 网易 QQ
          6 -> try {
            activity.getContent.launch(lrcMimeType) // 手动选择
          } catch (e: ActivityNotFoundException) {
            ToastUtil.show(activity, R.string.activity_not_found_tip)
          }

          7 -> MusicServiceRemote.service?.updateLyrics(IgnoredProvider) // 忽略
          8 -> TODO() // 调整字体大小
          9 -> activity.showLyricOffsetView() // 调整时间轴
        }
      }.show()
    // TODO: update LyricsFragment
  }

  companion object {
    private val lrcMimeType: String
      get() = MimeTypeMap.getSingleton().getMimeTypeFromExtension("lrc") ?: "*/*"
  }
}
